udffsck/udffsck.c

```diff
@@ -0,0 +1,3093 @@
+                    map_chunk(fd, dev, chunk, st_size); 
+
+                    memcpy((uint8_t *)(dev[chunk]+offset), fidArray+prevExtLength, ead->extLength);
+                    prevExtLength += ead->extLength;
+                    break;
+            }
+        }     
+    }
+
+    dbg("3 FID inspection copyback done.\n");
+    //free array
+    free(fidArray);
+    (*status) |= tempStatus;
+    return 0;
+}
+
+/**
+ * \brief FID parsing function
+ *
+ * This function pareses via FIDs. It continues to its FE using get_file() function. 
+ * Checks and fixes *Unique ID*, *Serial Numbers* or unfinished writings.
+ *
+ * This fucntion is complement to get_file() and translate_fid().
+ *
+ * \param[in,out] *dev memory mapped device
+ * \param[in] *disc udf_disc structure
+ * \param[in] lbnlsn LBN offset against LSN
+ * \param[in] lsn actual LSN
+ * \param[in] *base base pointer for for FID area
+ * \param[in,out] *pos actial position in FID area
+ * \param[in] *stats file system status
+ * \param[in] depth depth of FE for printing
+ * \param[in] *seq VDS sequence
+ * \param[out] *status run status
+ *
+ * \return 0 -- everything OK
+ * \return 1 -- Unknown descriptor found
+ * \return 252 -- FID checksum failed
+ * \return 251 -- FID CRC failed
+ */
+uint8_t inspect_fid(int fd, uint8_t **dev, const struct udf_disc *disc, size_t st_size, uint32_t lbnlsn, uint32_t lsn, uint8_t *base, uint32_t *pos, struct filesystemStats *stats, uint32_t depth, vds_sequence_t *seq, uint8_t *status) {
+    uint32_t flen, padding;
+    uint32_t lsnBase = lbnlsn; 
+    struct fileIdentDesc *fid = (struct fileIdentDesc *)(base + *pos);
+    struct fileInfo info = {0};
+    uint32_t offset = 0, chunk = 0;
+    uint64_t position = 0;
+    uint32_t chunksize = CHUNK_SIZE;
+
+    dbg("FID pos: 0x%x\n", *pos);
+    if (!checksum(fid->descTag)) {
+        err("[inspect fid] FID checksum failed.\n");
+        return -4;
+        warn("DISABLED ERROR RETURN\n");
+    }
+    if (le16_to_cpu(fid->descTag.tagIdent) == TAG_IDENT_FID) {
+        dwarn("FID found (%d)\n",*pos);
+        flen = 38 + le16_to_cpu(fid->lengthOfImpUse) + fid->lengthFileIdent;
+        padding = 4 * ((le16_to_cpu(fid->lengthOfImpUse) + fid->lengthFileIdent + 38 + 3)/4) - (le16_to_cpu(fid->lengthOfImpUse) + fid->lengthFileIdent + 38);
+
+        dbg("lengthOfImpUse: %d\n", fid->lengthOfImpUse);
+        dbg("flen+padding: %d\n", flen+padding);
+        if(crc(fid, flen + padding)) {
+            err("FID CRC failed.\n");
+            return -5;
+            warn("DISABLED ERROR RETURN\n");
+        }
+        dbg("FID: ImpUseLen: %d\n", fid->lengthOfImpUse);
+        dbg("FID: FilenameLen: %d\n", fid->lengthFileIdent);
+        if(fid->lengthFileIdent == 0) {
+            dbg("ROOT directory\n");
+        } else {
+            char *namebuf = calloc(1,256*2);
+            memset(namebuf, 0, 256*2);
+            int size = decode_utf8(fid->fileIdent, namebuf, fid->lengthFileIdent);
```

Comment:
In current master branch, this function returns size_t. And value `(size_t)-1` means that decoding failed.